### PR TITLE
add analysis page tabs

### PR DESCRIPTION
**Pull request by opadron**
Now, only one analysis page is rendered at a time.
This commit adds a new tab bar at the top of the
analysis UI.  Each tab in the tab bar represents
an analysis page.  User can switch between the
currently active page and another page by clicking
on its tab.

Analysis pages can now take on one of three
states: active - for when the page is the
currently active page, enabled - for other pages
that can be switched to, and disabled - for pages
that cannot be viewed at the moment, but may
become enabled as a result of some pending or
later operation.

Analysis pages can now register "tabEnter" and
"tabExit" analysis actions.  These actions will be
automatically triggered every time an analysis
page is newly enabled and when it has been
switched out for another page, respectively.

TODO: Modify the existing modules to work better
with the new "tab-based" way of doing things.
Need to do things like show all the pages up-front
(so the user can see all the tabs), and manage
page enabling/disabling as user navigates.

TODO: Tab bar is uuuuuuugly! Need a CSS wizard to
give it some love! ;P

  - Added new Redux state tree variables

    - analysis.currentPage

    - analysis.objects[\<pageId\>].enabled

  - Added new reducer action types

    - DISABLE_ANALYSIS_PAGE

    - ENABLE_ANALYSIS_PAGE

    - SET_CURRENT_ANALYSIS_PAGE

    - TOGGLE_ANALYSIS_PAGE

  - Added new global action creators

    - disableAnalysisPage

    - enableAnalysisPage

    - setCurrentAnalysisPage

    - toggleAnalysisPage

  - Reworked existing global action creators

    - triggerAnalysisAction can now take a
      "nothrow" option which, when set, will not
      throw an exception for a missing action.

    - truncateAnalysisPages can now take an
      options argument that allows developers to
      exercise more fine-grained control over its
      effect on targetted pages.  The recognized
      attributes are as follows:

        - clear: if set, targetted pages will be
          cleared of their elements. (default:
          true)

        - disable: if set, targetted pages will be
          disabled (default: true)

        - remove: if set, targetted pages will be
          removed, outright.  Setting this option
          renders the others moot.  (default:
          true)

  - Analysis pages can now register "tabEnter" and
    "tabExit" analysis actions, which will be
    called automatically.

  - Added new React component:
    body/analysis/tab-bar

  - Tweaked UI specs to have more tab-friendly
    names.

### CODE REVIEW

File: web_client/reducer/analysis.jsx [hunk 1]
```diff
@@ -412,13 +432,37 @@ const removeAnalysisPageByKey = (analysis, key) => {
...
+      }
+
+      if (clear) {
+        let idSet = Object.keys(
+          pagesToModify
+            .map((page) => objects[page].elements)
+            .reduce((a, b) => a.concat(b), [])
+            .map((element) => [element, true])
+            .reduce(objectReduce, {})
+        );
```
Comment (opadron) [inline web_client/reducer/analysis.jsx]: Note to self: This can be simplified.

**Comment by opadron** [comment]
Since it shouldn't affect integration, I went ahead and added markers to the `file-selection` elements that will appear for items that have a `sumoDataType` metadata attribute assigned to them.  This feature should satisfy the in-UI type requirements for the upcoming milestone.

**Comment by opadron** [comment]
@osumo/developers PTAL.  This, and #40 need review.

File: web_client/actions.jsx [hunk 2]
```diff
@@ -35,13 +35,23 @@ export const addAnalysisPage = (page) => promiseAction(
   (D, S) => {
```
Comment (waxlamp) [inline web_client/actions.jsx:35]: Might want better names for these function arguments. `D` and `S` are rather opaque, and variables in JavaScript that start with capital letters are supposed to be constructors/classnames.

File: web_client/actions.jsx [hunk 3]
```diff
@@ -80,6 +90,34 @@ export const closeDialog = (byRouter = false) => promiseAction(
   }
 );
 
+export const disableAnalysisPage = (page) => promiseAction(
+  (D, S) => {
```
Comment (waxlamp) [inline web_client/actions.jsx:94]: See above.

File: web_client/actions.jsx [hunk 4]
```diff
@@ -80,6 +90,34 @@ export const closeDialog = (byRouter = false) => promiseAction(
...
+    let { analysis } = S();
+    page = analysis.objects[id];
+    if (page) { page = { ...page }; }
+
+    return page;
+  }
+);
+
+export const enableAnalysisPage = (page) => promiseAction(
+  (D, S) => {
```
Comment (waxlamp) [inline web_client/actions.jsx:108]: See above

File: web_client/actions.jsx [hunk 5]
```diff
@@ -470,19 +499,39 @@ export const submitRegistrationForm = (form) => promiseAction(
   }
 );
 
+export const toggleAnalysisPage = (page) => promiseAction(
+  (D, S) => {
+    D({ type: ACTION_TYPES.TOGGLE_ANALYSIS_PAGE, page });
+
+    let { id } = page;
+    if (isUndefined(id)) { id = page; }
```
Comment (waxlamp) [inline web_client/actions.jsx:507]: Split this line up so the code block is in standard format.

File: web_client/actions.jsx [hunk 6]
```diff
@@ -470,19 +499,39 @@ export const submitRegistrationForm = (form) => promiseAction(
...
+    let { analysis } = S();
+    page = analysis.objects[id];
+    if (page) { page = { ...page }; }
```
Comment (waxlamp) [inline web_client/actions.jsx:510]: Split this line up so the code block is in standard format.

File: web_client/actions.jsx [hunk 7]
```diff
@@ -470,19 +499,39 @@ export const submitRegistrationForm = (form) => promiseAction(
...
+
+  nothrow = nothrow || false;
+  args = args || [];
+
+  return promiseAction(
     (D, S) => {
       let callback = globals.analysisActionTable[page.key];
       if (callback) { callback = callback[action]; }
       if (!callback) {
+        if (nothrow) { return null; }
```
Comment (waxlamp) [inline web_client/actions.jsx:534]: Split this line up so the code block is in standard format.

File: web_client/components/body/analysis/tab-bar.jsx [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    return (
+      <div className='analysis-tab-bar'>
+        { tabElements }
+      </div>
+    );
+  }
+}
+
+export default AnalysisTabBar;
+
```
Comment (waxlamp) [inline web_client/components/body/analysis/tab-bar.jsx:65]: Remove blank line

**Review by waxlamp — APPROVED** [review]
Approving this, despite the minor style suggestions I left as comments. Functionality looks great!

**Comment by opadron** [comment]
Approved.  See #42 for integration.

**Pull request closed without merging**